### PR TITLE
Cleaned ho3

### PR DESCRIPTION
**Pull request by Klopfe**

### CODE REVIEW

File: sparse_ho/ho.py [hunk 1]
```diff
@@ -13,34 +13,41 @@ def grad_search(
 
     Parameters
     ----------
-    algo: TODO
-    criterion: TODO
-    model: TODO
+    algo: BaseAlgo instance
```
Comment (mathurinm) [inline sparse_ho/ho.py:16]: harmonize: instance of ...

File: sparse_ho/ho.py [hunk 2]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        algorithm used to compute hypergradient.
+    criterion: BaseCriterion
```
Comment (mathurinm) [inline sparse_ho/ho.py:18]: ditto

File: sparse_ho/ho.py [hunk 3]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        criterion to optimize during hyperparameter optimization
+        (outer optimization problem).
+    model: BaseModel
```
Comment (mathurinm) [inline sparse_ho/ho.py:21]: ditto

File: sparse_ho/ho.py [hunk 4]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        model on which hyperparameter has to be selected
+        (inner optimization problem).
+    X: {ndarray, sparse matrix} of (n_samples, n_features)
```
Comment (mathurinm) [inline sparse_ho/ho.py:24]: array like, shape (n_samples, n_features)

File: sparse_ho/ho.py [hunk 5]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        Data.
+    y: {ndarray, sparse matrix} of shape (n_samples,)
```
Comment (mathurinm) [inline sparse_ho/ho.py:26]: do you need sparse y ?

File: sparse_ho/ho.py [hunk 6]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+    algo: BaseAlgo instance
+        algorithm used to compute hypergradient.
+    criterion: BaseCriterion
+        criterion to optimize during hyperparameter optimization
+        (outer optimization problem).
+    model: BaseModel
+        model on which hyperparameter has to be selected
+        (inner optimization problem).
+    X: {ndarray, sparse matrix} of (n_samples, n_features)
+        Data.
```
Comment (mathurinm) [inline sparse_ho/ho.py:25]: Design matrix is more explicit IMO as y is data too

File: sparse_ho/ho.py [hunk 7]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+    y: {ndarray, sparse matrix} of shape (n_samples,)
+        Target.
     log_alpha0: float
-        log of the regularization coefficient alpha
-    tol : float
-        tolerance for the inner optimization solver
-    monitor: Monitor object
-        used to store the value of the cross-validation function
+        log of the regularization coefficient alpha.
```
Comment (mathurinm) [inline sparse_ho/ho.py:29]: initial value of ?

File: sparse_ho/ho.py [hunk 8]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+    monitor: instance of Monitor
+        used to store the value of the cross-validation function.
```
Comment (mathurinm) [inline sparse_ho/ho.py:31]: it can do more

File: sparse_ho/ho.py [hunk 9]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
     n_outer: int
```
Comment (mathurinm) [inline sparse_ho/ho.py:32]: write: `n_outer: int, optional (default=100)`

File: sparse_ho/ho.py [hunk 10]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
-        number of maximum iteration in the outer loop (for the line search)
+        Defaults to 100.
+        number of maximum iteration in the outer loop (for the line search).
```
Comment (mathurinm) [inline sparse_ho/ho.py:34]: iterations
number of maximum line search iterations ? Or number of updates of alpha?

File: sparse_ho/ho.py [hunk 11]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+    verbose: Bool
```
Comment (mathurinm) [inline sparse_ho/ho.py:35]: bool, optional (default=False)

File: sparse_ho/ho.py [hunk 12]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        Defaults False. Indicates whether information about hyperparameter
+        optimization process  is printed or not.
```
Comment (mathurinm) [inline sparse_ho/ho.py:37]: two spaces

File: sparse_ho/ho.py [hunk 13]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
     tolerance_decrease: string
-        tolerance decrease strategy for approximate gradient
-    criterion: string
-        criterion to optimize during hyperparameter optimization
-        you may choose between "cv" and "sure"
-    random_state: int
-    beta_star: np.array, shape (n_features,)
-        True coefficients of the underlying model (if known)
-        used to compute metrics
+        Defaults "constant". Tolerance decrease strategy for approximate gradient.
```
Comment (mathurinm) [inline sparse_ho/ho.py:39]: optional, etc

File: sparse_ho/ho.py [hunk 14]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+    tol : float
```
Comment (mathurinm) [inline sparse_ho/ho.py:40]: extra space

File: sparse_ho/ho.py [hunk 15]
```diff
@@ -13,34 +13,41 @@ def grad_search(
...
+        Defaults 1e-5. Tolerance for the inner optimization solver.
+    t_max: float
+        Defaults 10000. Maximum running time threshold.
```
Comment (mathurinm) [inline sparse_ho/ho.py:43]: in seconds ?

File: sparse_ho/ho.py [hunk 16]
```diff
@@ -54,26 +61,17 @@ def _grad_search(
...
-    Parameters
-    --------------
-    TODO
-
-    """
 
     try:
-        lambdak = log_alpha0.copy()
-        old_lambdak = lambdak.copy()
+        log_alphak = log_alpha0.copy()
```
Comment (mathurinm) [inline sparse_ho/ho.py:66]: this try except can be improved IMO. It's not obvious than you're making an if based on if there's a single parameter alpha or a full array

File: sparse_ho/ho.py [hunk 17]
```diff
@@ -86,67 +84,67 @@ def _grad_search(
             old_tol = seq_tol[i - 1]
         except Exception:
             old_tol = seq_tol[0]
-        g_func, grad_lambda = _get_val_grad(lambdak, tol=tol, monitor=monitor)
+        val_outer, grad_outer = _get_val_grad(log_alphak, tol=tol, monitor=monitor)
 
-        old_grads.append(norm(grad_lambda))
+        old_grads.append(norm(grad_outer))
```
Comment (mathurinm) [inline sparse_ho/ho.py:89]: if it holds norm of gradients I would call it grad_norms, old grads make me think it holds full gradients

File: sparse_ho/ho.py [hunk 18]
```diff
@@ -86,67 +84,67 @@ def _grad_search(
...
             print("Nan present in gradient")
             break
 
-        if L_lambda is None:
+        if L_log_alpha is None:
             if old_grads[-1] > 1e-3:
                 # make sure we are not selecting a step size that is too small
                 try:
-                    L_lambda = old_grads[-1] / np.sqrt(len(lambdak))
+                    L_log_alpha = old_grads[-1] / np.sqrt(len(log_alphak))
```
Comment (mathurinm) [inline sparse_ho/ho.py:98]: same try except

File: sparse_ho/ho.py [hunk 19]
```diff
@@ -6,41 +6,47 @@
...
-    tol : float
-        tolerance for the inner optimization solver
-    monitor: Monitor object
-        used to store the value of the cross-validation function
-    n_outer: int
-        number of maximum iteration in the outer loop (for the line search)
-    tolerance_decrease: string
-        tolerance decrease strategy for approximate gradient
-    criterion: string
+    algo: instance of BaseAlgo
```
Comment (agramfort) [inline sparse_ho/ho.py:16]: guys these classes do not actually exist !

see my attempt to solve this : https://github.com/QB3/sparse-ho/pull/52

this still needs to be fixed with a BaseAlgo and your API for criterion is not consistent as revealed by the failing tests.

**Pull request merged**